File: typescript-connector/src/models/Offer.ts
// Class from DFC Business Ontology: #Offer
import { SemanticObject } from "../core/SemanticObject.js";
import type { CatalogItem } from "./CatalogItem.js";
import type { CustomerCategory } from "./CustomerCategory.js";

export interface OfferParams {
  discount?: number;
  stockLimitation?: number;
  concernedBy?: string;
  hasPrice?: string;
  listedIn?: string;
  date?: string;
  description?: string;
  name?: string;
  characteristicOf?: string;
  hasDimension?: string;
  offers?: (CatalogItem | string)[];
  offersTo?: (CustomerCategory | string)[];
}
export class Offer extends SemanticObject {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:Offer";
  }

  discount?: number;
  stockLimitation?: number;
  concernedBy?: string;
  hasPrice?: string;
  listedIn?: string;
  date?: string;
  description?: string;
  name?: string;
  characteristicOf?: string;
  hasDimension?: string;
  offers?: (CatalogItem | string)[];
  offersTo?: (CustomerCategory | string)[];

  constructor(
    semanticId: string,
    params?: OfferParams,
  ) {
    super(semanticId);
    this.discount = params?.discount;
    this.stockLimitation = params?.stockLimitation;
    this.concernedBy = params?.concernedBy;
    this.hasPrice = params?.hasPrice;
    this.listedIn = params?.listedIn;
    this.date = params?.date;
    this.description = params?.description;
    this.name = params?.name;
    this.characteristicOf = params?.characteristicOf;
    this.hasDimension = params?.hasDimension;
    this.offers = params?.offers;
    this.offersTo = params?.offersTo;
    this.semanticType = Offer.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:discount", () => this.discount);
    this.registerSemanticProperty("dfc-b:stockLimitation", () => this.stockLimitation);
    this.registerSemanticProperty("dfc-b:concernedBy", () => this.concernedBy);
    this.registerSemanticProperty("dfc-b:hasPrice", () => this.hasPrice);
    this.registerSemanticProperty("dfc-b:listedIn", () => this.listedIn);
    this.registerSemanticProperty("dfc-b:date", () => this.date);
    this.registerSemanticProperty("dfc-b:description", () => this.description);
    this.registerSemanticProperty("dfc-b:name", () => this.name);
    this.registerSemanticProperty("dfc-b:characteristicOf", () => this.characteristicOf);
    this.registerSemanticProperty("dfc-b:hasDimension", () => this.hasDimension);
    this.registerSemanticProperty("dfc-b:offers", () => this.offers);
    this.registerSemanticProperty("dfc-b:offersTo", () => this.offersTo);
  }
  static {
    SemanticObject.typeRegistry.set(Offer.SEMANTIC_TYPE, Offer);
  }
}

File: ruby-gem/lib/models/offer.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #Offer
require_relative '../core/semantic_object'


module DfcLinkmlConnector
  module Models
    class Offer < Core::SemanticObject
      SEMANTIC_TYPE = "dfc-b:Offer".freeze

      # @return [Float]
      attr_accessor :discount

      # @return [Float]
      attr_accessor :stock_limitation

      # @return [String]
      attr_accessor :concerned_by

      # @return [String]
      attr_accessor :price

      # @return [String]
      attr_accessor :listed_in

      # @return [String]
      attr_accessor :offers

      # @return [String]
      attr_accessor :offers_to

      # @return [String]
      attr_accessor :date

      # @return [String]
      attr_accessor :description

      # @return [String]
      attr_accessor :name

      # @return [String]
      attr_accessor :characteristic_of

      # @return [String]
      attr_accessor :dimension

      # @param semanticId [String]
      # @param discount: nil, stockLimitation: nil, concernedBy: nil, price: nil, listedIn: nil, offers: nil, offersTo: nil, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil
      def initialize(semanticId, discount: nil, stockLimitation: nil, concernedBy: nil, price: nil, listedIn: nil, offers: nil, offersTo: nil, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil)
        super(semanticId)
        @discount = discount
        @stock_limitation = stockLimitation
        @concerned_by = concernedBy
        @price = price
        @listed_in = listedIn
        @offers = offers
        @offers_to = offersTo
        @date = date
        @description = description
        @name = name
        @characteristic_of = characteristicOf
        @dimension = dimension
        self.semanticType = "dfc-b:Offer"
        registerSemanticProperty("dfc-b:discount", &method("discount")).valueSetter = method("discount=")
        registerSemanticProperty("dfc-b:stockLimitation", &method("stock_limitation")).valueSetter = method("stock_limitation=")
        registerSemanticProperty("dfc-b:concernedBy", &method("concerned_by")).valueSetter = method("concerned_by=")
        registerSemanticProperty("dfc-b:hasPrice", &method("price")).valueSetter = method("price=")
        registerSemanticProperty("dfc-b:listedIn", &method("listed_in")).valueSetter = method("listed_in=")
        registerSemanticProperty("dfc-b:offers", &method("offers")).valueSetter = method("offers=")
        registerSemanticProperty("dfc-b:offersTo", &method("offers_to")).valueSetter = method("offers_to=")
        registerSemanticProperty("dfc-b:date", &method("date")).valueSetter = method("date=")
        registerSemanticProperty("dfc-b:description", &method("description")).valueSetter = method("description=")
        registerSemanticProperty("dfc-b:name", &method("name")).valueSetter = method("name=")
        registerSemanticProperty("dfc-b:characteristicOf", &method("characteristic_of")).valueSetter = method("characteristic_of=")
        registerSemanticProperty("dfc-b:hasDimension", &method("dimension")).valueSetter = method("dimension=")
      end

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end
